server/channels/app/password/hashers/hashers.go

```diff
@@ -0,0 +1,122 @@
+//
+//		switch phc.Id {
+//	  - case PBKDF2FunctionId:
+//	  - case NewHasherFunctionId:
+//	    return LatestHasher, phc
+//	  - case PBKDF2FunctionId:
+//	  - return DefaultPBKDF2()
+//	    // If the function ID is unknown, return the default hasher
+//	    default:
+//	    return getDefaultHasher(phcString)
+//
+// ```
+//
+// Note that the migration happens in [users.MigratePassword], which is triggered
+// whenever the user enters their password and an old hashing method is
+// identified when parsing their stored hashed password: the
+// [users.OutdatedPasswordHashingError] is thrown by [users.CheckUserPassword].
+// This means that the older password hashers can *never* be removed, unless all
+// users whose passwords are not migrated are either forced to re-login, or
+// forced to generate a new password.
+package hashers
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/mattermost/mattermost/server/v8/channels/app/password/parser"
+)
+
+// PasswordHasher is a password hasher compliant with the PHC string format:
+// https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md
+//
+// Implementations of this interface need to make sure that all of the methods
+// are thread-safe.
+type PasswordHasher interface {
+	// Hash computes a hash of the provided password, returning a PHC string
+	// containing all the information that was needed to compute it: the function
+	// used to generate it, its version, parameters and salt, if needed.
+	//
+	// If an implementation of Hash needs a salt to generate the hash, it will
+	// create one, so that callers don't need to provide it.
+	Hash(password string) (string, error)
+
+	// CompareHashAndPassword compares the parsed PHC and a provided password,
+	// returning an error if and only if the hashes do not match.
+	//
+	// Implementations need to make sure that the comparisons are done in constant
+	// time; e.g., using crypto/internal/fips140/subtle.ConstantTimeCompare.
+	CompareHashAndPassword(hash parser.PHC, password string) error
+}
+
+const (
+	// Maximum password length for all password hashers
+	PasswordMaxLengthBytes = 72
+)
+
+var (
+	// LatestHasher is the hasher currently in use.
+	// Any password hashed with a different hasher must be migrated to this one.
+	LatestHasher PasswordHasher = DefaultPBKDF2()
+
+	// ErrPasswordTooLong is the error returned when the provided password is
+	// longer than [PasswordMaxLengthBytes].
+	ErrPasswordTooLong = fmt.Errorf("password too long; maximum length in bytes: %d", PasswordMaxLengthBytes)
+
+	// ErrMismatchedHashAndPassword is the error returned when the provided
+	// password does not match the stored hash
+	ErrMismatchedHashAndPassword = fmt.Errorf("hash and password do not match")
+)
+
+// getDefaultHasher returns a [BCrypt] hasher, which is somewhat special, since it
+// uses the whole PHC string as the hash.
+func getDefaultHasher(phcString string) (PasswordHasher, parser.PHC) {
+	return NewBCrypt(), parser.PHC{Hash: phcString}
```

Comment:
Why is BCrypt the default? Or do we mean something like `getOriginalHasher`?

Reply:
Default as in "if we don't identify what hasher it is because it's not PHC-based, choose bcrypt". Your naming sounds good as well :)